FIX Ensure dtype of categories is `object` for strings in `OneHotEncoder`

betatim opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Closes #25171

#### What does this implement/fix? Explain your changes.
This makes it so that the `enc.categories_` attribute of `OneHotEncoder` is contains an array of dtype `object` when using predefined categories that are strings. This makes it consistent with the dtype of `enc.categories_` when the categories are determine during `fit`. In general to compare a sequence of bytes to a string you need to assume an encoding, otherwise you can't really compare them. But I don't understand enough about the numpy type system to know if it would take care of this already? Ideas an input welcome.

betatim committed:
Ensure dtype of categories is object for strings

adrinjalali commented:
Why do we not want to have `<UXXX` types?

```py
>>> np.asarray(["test"]).dtype
dtype('<U4')
```

betatim commented:
> Why do we not want to have `<UXXX` types?

I assumed that there is a reason why when you do [something slightly different but similar](https://github.com/scikit-learn/scikit-learn/issues/25171#issuecomment-1346200301) (don't provide categories to constructor, rely on fit to find them) you end of with `object` as dtype. But I've not (yet) found a comment explaining why the code does what it does :-/

adrinjalali commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -97,7 +97,12 @@ def _fit(
                     cats = result
             else:
-                cats = np.array(self.categories[i], dtype=Xi.dtype)
+                if np.issubdtype(Xi.dtype, np.str_):
+                    Xi_dtype = object
+                else:
+                    Xi_dtype = Xi.dtype
```
What wold be nice is to have a way to find a "common dtype" where it could hold both data from categories and X I think.

betatim replied:
I tried `np.result_type()` which would do that. Didn't go with it in the end because there seems to be this special casing of "string like numpy? then object!" in `_check_X`. Any idea how we could find out what the rationale behind that is?

adrinjalali replied:
yeah the code is moved around quite a bit, can't tell what the origin is. cc'ing @thomasjpfan here.

I think `result_type` should be a good enough option though.

ogrisel replied:
For string values of categorical variables, I think it's better to always use the object dtype (the only available representation for variable length strings stored in numpy arrays) .

adrinjalali replied:
hopefully my worry about performance regression here is unfounded 😝 

I feel like we're adding a bunch of pointer lookups here. There's a big difference between object dtypes and other dtypes in numpy. Object dtypes can't even be persisted with numpy format.

ogrisel replied:
The arrays in the `.categories_` are small. And most of the time we access such an attribute we want to manipulate `str` objects in the end (e.g. to display category labels for a pipeline inspection tool.

ogrisel commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -97,7 +97,12 @@ def _fit(
                     cats = result
             else:
-                cats = np.array(self.categories[i], dtype=Xi.dtype)
+                if np.issubdtype(Xi.dtype, np.str_):
+                    Xi_dtype = object
```
```suggestion
                if np.issubdtype(Xi.dtype, np.str_):
                    # Always convert string categories to objects to avoid
                    # unexpected string truncation for longer category labels
                    # passed in the constructor.
                    Xi_dtype = object
```

ogrisel commented on sklearn/preprocessing/tests/test_encoders.py:
```diff
@@ -1939,3 +1939,18 @@ def test_ordinal_set_output():
+    enc.fit([["as", "1"]])
+
+    for cat in enc.categories_:
+        assert cat.dtype == object
```
Could you also please check that the values in `enc.categories_` are not truncated?

ogrisel reviewed (commented):
Please also add an entry in the changelog, maybe for 1.2.1.

Once the following is addressed, LGTM:

betatim committed:
Check for truncated category labels

betatim committed:
Remove test with bytes and unicode categories

It isn't obvious how to compare a sequence of bytes to a string, without
knowing what encoding to use. Removing this test as it fails now that we
use object as dtype for strings.

betatim commented on sklearn/preprocessing/tests/test_encoders.py:
```diff
@@ -1346,7 +1346,7 @@ def test_one_hot_encoder_sparse_deprecated():
 # deliberately omit 'OS' as an invalid combo
 @pytest.mark.parametrize(
-    "input_dtype, category_dtype", ["OO", "OU", "UO", "UU", "US", "SO", "SU", "SS"]
+    "input_dtype, category_dtype", ["OO", "OU", "UO", "UU", "SO", "SU", "SS"]
```
The combination of `fit` input with type `U` and predefined categories as `S` lead to comparing an array of bytes to an array of strings.

In `_fit` we ended up with `cats = array([b'b', b'a'], dtype=object)` and `Xi = array(['b', 'a'], dtype='<U1')`, which then failed the `_check_unknown()` test. I think it makes sense that this happens. An array of bytes and an array of strings can't really be compared, without assuming an encoding.

A bit unsure what to do here. I don't like having to modify an existing test to "make it pass".

betatim replied:
I've added a test for this specific case, as well as code in the estimator that raises a particular message to help users figure out what happened. WDYT?

glemaitre replied:
+1 on my side.

betatim commented:
I'll add a change log entry later.

For now I have two questions: (1) about the modified test (see comment) and (2) why is the dtype taken from `X` that is passed to `fit` instead of from `self.categories`, when the user explicitly specified the known categories. My instinct would have been to take the type from the specified category and require the input to `fit` to match/be compatible.

adrinjalali commented:
> My instinct would have been to take the type from the specified category and require the input to fit to match/be compatible.

I agree with them having to be compatible, but they don't have to be the same.

betatim committed:
Add explicit test for mixed string-bytes categories

betatim committed:
Add explicit error message for incompatible category values

A predefined category of bytes can not be compared to input data that is
a unicode string.

betatim committed:
Fix typo

betatim committed:
Fix tests

betatim commented on sklearn/preprocessing/tests/test_encoders.py:
```diff
@@ -1376,6 +1376,22 @@ def test_encoders_string_categories(input_dtype, category_dtype, array_type):
+    categories = [np.array(["b", "a"], dtype="S")]
+    ohe = OneHotEncoder(categories=categories, sparse_output=False)
+
+    with pytest.raises(ValueError, match="Found unknown categories"):
```
I've now switched to "Found incompatible types between values and predefined categories in column 0.". I don't think we could raise the error much earlier than where it is now.

betatim commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -97,7 +97,15 @@ def _fit(
+                else:
+                    Xi_dtype = Xi.dtype
+
+                cats = np.array(self.categories[i], dtype=Xi_dtype)
```
I don't know :-/ so I'd leave it as it is as this isn't new code, it just moved down a few lines.

adrinjalali commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -97,7 +97,26 @@ def _fit(
+                    and Xi.dtype.kind != "S"
+                ):
+                    msg = (
+                        "Found incompatible types between values and predefined"
+                        " categories in column {0}.".format(i)
```
would be nice to print what those types are.

adrinjalali reviewed (commented):
Otherwise LGTM.

Needs a changelog though.

betatim committed:
Add what's new entry

betatim committed:
Improve error message by outputing type names

betatim committed:
Merge branch 'main' into fix-categories-dtype

glemaitre commented on doc/whats_new/v1.3.rst:
```diff
@@ -52,6 +55,7 @@ Changelog
   :pr:`24935` by :user:`Seladus <seladus>`, :user:`Guillaume Lemaitre <glemaitre>`, and
   :user:`Dea María Léon <deamarialeon>`.
 
+
```
```suggestion
```

glemaitre commented on sklearn/preprocessing/tests/test_encoders.py:
```diff
@@ -1376,6 +1376,28 @@ def test_encoders_string_categories(input_dtype, category_dtype, array_type):
+    with pytest.raises(
+        ValueError,
+        match=(
+            "Found incompatible types ('bytes' vs 'str_') between values and predefined"
```
Could you declare this string just before the context manager. It will be easier to read.

glemaitre reviewed (approved):
Just nitpicking.

betatim committed:
Remove blank line

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

betatim committed:
Move error message

thomasjpfan commented on sklearn/preprocessing/tests/test_encoders.py:
```diff
@@ -1376,6 +1376,27 @@ def test_encoders_string_categories(input_dtype, category_dtype, array_type):
+    categories = [np.array(["b", "a"], dtype="S")]
+    ohe = OneHotEncoder(categories=categories, sparse_output=False)
+
+    msg = (
```
To fix the failing CI:

```suggestion
    msg = re.escape(
```

thomasjpfan commented on sklearn/preprocessing/tests/test_encoders.py:
```diff
@@ -1376,6 +1376,27 @@ def test_encoders_string_categories(input_dtype, category_dtype, array_type):
     assert_array_equal(X_trans, expected)
 
 
+def test_mixed_string_bytes_categoricals():
```
Small typo:

```suggestion
def test_mixed_string_bytes_categoricals():
```

betatim replied:
What is the typo?

thomasjpfan replied:
Ah this is my mistake. There is no typo. 😅

thomasjpfan commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -97,7 +97,27 @@ def _fit(
+                    and isinstance(cats[0], bytes)
+                    and Xi.dtype.kind != "S"
+                ):
+                    msg = (
+                        f"Found incompatible types ('{type(cats[0]).__name__}' vs"
+                        f" '{type(Xi[0]).__name__}') between values and predefined"
+                        f" categories in column {i}."
+                    )
```
I think this is a slightly clearer error message. `'bytes'` can be hard coded because of the `if` statement above it:

```suggestion
                    msg = (
                        f"In column {i}, the predefined categories has type"
                        " 'bytes' which is incompatible with"
                        f" values of type '{type(Xi[0]).__name__}'."
                    )
                    raise ValueError(msg)
```

Although, using bytes as predefined categories is an edge case, but this PR is a backward incompatible change. If we are okay with it, then it needs to be in the "Changed Model" section of the whats new.

betatim replied:
I prefer your message. Much less clunky than what I came up with. Also moved the change log entry.

thomasjpfan reviewed (commented):
Thank you for the PR!

betatim committed:
Add regex escape

betatim committed:
Switch to easier to understand error message

betatim committed:
Move change log entry as it is a breaking change

thomasjpfan reviewed (approved):
Minor comment, otherwise LGTM

betatim committed:
Update doc/whats_new/v1.3.rst

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>
